internal/fs/fs.go

```diff
@@ -2148,6 +2154,13 @@ func (fs *fileSystem) OpenDir(
 	fs.handles[handleID] = handle.NewDirHandle(in, fs.implicitDirs)
 	op.Handle = handleID
 
+	// Invalidates the kernel dir-cache once the last cached response is out of
+	// kernelDirCacheTTL.
+	op.KeepCache = !in.ShouldInvalidateKernelDirCache(fs.kernelDirCacheTTL)
```

Comment:
this operation should be guarded by in.Lock() ?

Reply:
Looks like it or it might be affected by a concurrent ReadDir call. An RLock() should suffice, if we have it. If we don't have it in the DirInode interface, then we can do it within the ShouldInvalidateKernelDirCache method itself.